### PR TITLE
Item 6848: Sample Type Designer Update

### PR DESCRIPTION
**Pull request by labkey-tchad**

### CODE REVIEW

**Review by Sigmonia — APPROVED** [review]

File: src/org/labkey/test/components/labkey/ui/samples/SampleTypeDesigner.java [hunk 1]
```diff
@@ -0,0 +1,225 @@
...
+        {
+            if (inputs.get(i).get().equals(parentAlias))
+            {
+                aliasIndex = i;
+                break;
+            }
+        }
+        if (aliasIndex < 0)
+        {
+            throw new NotFoundException("No such parent alias: " + parentAlias);
```
Comment (labkey-danield) [inline src/org/labkey/test/components/labkey/ui/samples/SampleTypeDesigner.java]: Is it really a fatal issue if a user tries to remove a parentAlias that isn't there?

File: src/org/labkey/test/components/labkey/ui/samples/SampleTypeDesigner.java [hunk 2]
```diff
@@ -0,0 +1,225 @@
...
+    public SampleTypeDesigner setParentAlias(int index, @Nullable String alias, @Nullable String optionDisplayText)
+    {
+        elementCache().parentAlias(index).setValue(alias);
+        if (optionDisplayText != null)
+        {
+            elementCache().parentAliasSelect(index).select(optionDisplayText);
+        }
+        return this;
+    }
+
```
Comment (labkey-danield) [inline src/org/labkey/test/components/labkey/ui/samples/SampleTypeDesigner.java:181]: Would it be useful to have a setParetnAlias that uses the alias as the identifier? That is setParetnAlias(String alias, @Nullable optionDisplayText).

File: src/org/labkey/test/components/labkey/ui/samples/SampleTypeDesigner.java [hunk 3]
```diff
@@ -0,0 +1,225 @@
...
+    public String getParentAliasSelectText(int index)
+    {
+        return elementCache().parentAliasSelect(index).getSelections().get(0);
+    }
+
+    protected class ElementCache extends Component<?>.ElementCache
+    {
+        protected final Input nameInput = Input.Input(Locator.id("entity-name"), getDriver()).findWhenNeeded(this);
+        protected final Input nameExpressionInput = Input.Input(Locator.id("entity-nameExpression"), getDriver()).waitFor(this);
```
Comment (labkey-danield) [inline src/org/labkey/test/components/labkey/ui/samples/SampleTypeDesigner.java:202]: Why is this a waitFor and not a findWhenNeeded?

File: src/org/labkey/test/params/experiment/SampleSetDefinition.java [hunk 4]
```diff
@@ -75,20 +79,70 @@ public SampleSetDefinition addField(@NotNull FieldDefinition field)
...
+    protected String getKind()
+    {
+        return "SampleSet";
+    }
+
+    @NotNull
+    @Override
+    protected Map<String, Object> getOptions()
+    {
+        Map<String, Object> json = new HashMap<>();
```
Comment (labkey-danield) [inline src/org/labkey/test/params/experiment/SampleSetDefinition.java]: Kind of an interesting variable name...

File: src/org/labkey/test/BootstrapLocators.java [hunk 5]
```diff
@@ -2,10 +2,10 @@
...
 public abstract class BootstrapLocators
 {
-    public static final Locator infoBanner = Locator.tagWithClass("div", BannerType.INFO.getCss());
-    public static final Locator successBanner = Locator.tagWithClass("div", BannerType.SUCCESS.getCss());
-    public static final Locator errorBanner = Locator.tagWithClass("div", BannerType.ERROR.getCss());
-    public static final Locator warningBanner = Locator.tagWithClass("div", BannerType.WARNING.getCss());
+    public static final Locator.XPathLocator infoBanner = Locator.tagWithClass("div", BannerType.INFO.getCss());
+    public static final Locator.XPathLocator successBanner = Locator.tagWithClass("div", BannerType.SUCCESS.getCss());
+    public static final Locator.XPathLocator errorBanner = Locator.tagWithClass("div", BannerType.ERROR.getCss());
+    public static final Locator.XPathLocator warningBanner = Locator.tagWithClass("div", BannerType.WARNING.getCss());
```
Comment (labkey-danield) [inline src/org/labkey/test/BootstrapLocators.java]: Not really sure what this change has to to do with this story.

File: src/org/labkey/test/tests/SampleSetTest.java [hunk 6]
```diff
@@ -914,9 +886,7 @@ public void testDeleteSamplesSomeWithAssayData()
         portalHelper.addWebPart("Assay List");
         SampleSetHelper sampleHelper = new SampleSetHelper(this);
         log("Create a sample set");
-        sampleHelper.createSampleSet(SAMPLE_SET_NAME, null,
-                null,
-                sampleData);
+        sampleHelper.createSampleSet(new SampleSetDefinition(SAMPLE_SET_NAME), sampleData);
 
 //  Note that we currently will not find runs where the batch id references a sampleId.  See Issue 37918.
```
Comment (labkey-danield) [inline src/org/labkey/test/tests/SampleSetTest.java:891]: This comment should be removed, issue 37918 has been closed.

**Review by labkey-danield — APPROVED** [review]
I'm going to follow dev's lead and approve even though I have comments.
Looks ok, just a slight concern about an assert for a situation that is not a catastrophic.

File: src/org/labkey/test/components/labkey/ui/samples/SampleTypeDesigner.java (hunk 1 above)
Comment (labkey-tchad) [reply src/org/labkey/test/components/labkey/ui/samples/SampleTypeDesigner.java]: I prefer to have things throw than to no-op silently.

File: src/org/labkey/test/components/labkey/ui/samples/SampleTypeDesigner.java (hunk 3 above)
Comment (labkey-tchad) [reply src/org/labkey/test/components/labkey/ui/samples/SampleTypeDesigner.java:202]: Not sure. Much of this class was extracted from the `SampleSetDesignPage` in Sample Manager. I'm guessing this acts as insurance that the form has loaded so I'm hesitant to change it.

File: src/org/labkey/test/params/experiment/SampleSetDefinition.java (hunk 4 above)
Comment (labkey-tchad) [reply src/org/labkey/test/params/experiment/SampleSetDefinition.java]: Quite. This was a `JSONObject` in an earlier iteration. Changing it to `options`

File: src/org/labkey/test/BootstrapLocators.java (hunk 5 above)
Comment (labkey-tchad) [reply src/org/labkey/test/BootstrapLocators.java]: I think an earlier iteration needed access to `XPathLocator`-specific methods. Reverted.

**Comment by labkey-tchad** [comment]
A little background in case others want to review (@labkey-sweta @labkey-chrisj).

Creating/editing sample sets used to be two separate pages; one for the name, name expression, and import aliases; and one for defining the fields. APIs, and test helpers were similarly divided. This merges tests' handling of sample set designs and sample set domains.

UI Related changes:
 - I refactored those methods to be more atomic and made the various `createSampleSet` methods take a `SampleSetDefinition` rather than having numerous overloads for various use cases. 
 - Created new `SampleTypeDesigner` component based on existing Sample Manager test helper.

API Related changes:
Similar to the UI, creating a Sample Set via API went from two calls ('createSampleSet' and 'saveDomain') to a single call ('createDomain').
 - `SampleSetAPIHelper` was updated to use `CreateDomainCommand` and utilize the improved functionality of `SampleSetDefinition` (see below)
 - Improvements to `FieldDefinition`
    - Extend `PropertyDescriptor` from remoteapi and renamed some methods to match. This allows you to use `FieldDefinition` directly with various remote API methods that take `PropertyDescriptor`.
    - Extend properties currently supported by the Java API (See `FieldDefinition.toJSONObject`).
    - Added test API support for special `User`, `Sample`, and `Flag` column types.
 - Add general interface for creating domain definitions via API (`DomainProps`)
    - `SampleSetDefinition` extends it and is widely used.
    - `ListDefinition` is just a proof of concept. Needs to be iterated on and integrated into `ListHelper` and a future `ListApiHelper`.

File: src/org/labkey/test/components/labkey/ui/samples/SampleTypeDesigner.java (hunk 2 above)
Comment (labkey-tchad) [reply src/org/labkey/test/components/labkey/ui/samples/SampleTypeDesigner.java:181]: Probably. Added.

**Review by labkey-chrisj — APPROVED** [review]

**Pull request merged**